### ISSUE TITLE
Reward for Protos has some problem

### ISSUE THREAD

**Issue by hepengli** [oxwhirl/smac#72]
I found that the reward for the race of Protos may not be correct. The problem is that when the enemy regenerates shield, the allies will receive rewards. The location of the problem is in the file "./env/starcraft2/starcraft2.py", line 729:

   728     if self.reward_only_positive:
   729         reward = abs(delta_enemy + delta_deaths)  # shield regeneration
   730     else:
   731         reward = delta_enemy + delta_deaths - delta_ally

By default, the argument self.reward_only_positive == True, and the reward should be always positive. To achieve this, line 729 uses abs() to convert "delta_enemy + delta_deaths" to a positive number. However, this will result in a mistake because when the Protos enemy units regenerate shield, the variable "delta_enemy" is negative. In this case, the allies should be punished. Nevertheless, the code in 729 will lead to the allies being rewarded.

**Comment by douglasrizzo** [comment]
Just to be clear, the problem you mention happens when `delta_enemy` is negative with a larger magnitude than `delta_deaths`, which would mean that agents are rewarded by injuring enemies, while not killing them. In this scenario, `delta_enemy + delta_deaths` would be negative, but the absolute of that would be positive.

Since `self.reward_only_positive` means we want to ignore negative rewards, I believe the best solution is to take the maximum of 0 and `delta_enemy + delta_deaths`. It would behave more closely like the reward when `self.reward_only_positive == False`, but never being negative and ignoring ally damage and deaths.

```python
if self.reward_only_positive:
    reward = max(0, delta_enemy + delta_deaths)  # shield regeneration
else:
    reward = delta_enemy + delta_deaths - delta_ally
```

**Comment by hepengli** [comment]
> Just to be clear, the problem you mention happens when `delta_enemy` is negative with a larger magnitude than `delta_deaths`, which would mean that agents are rewarded by injuring enemies, while not killing them. In this scenario, `delta_enemy + delta_deaths` would be negative, but the absolute of that would be positive.
> 
> Since `self.reward_only_positive` means we want to ignore negative rewards, I believe the best solution is to take the maximum of 0 and `delta_enemy + delta_deaths`. It would behave more closely like the reward when `self.reward_only_positive == False`, but never being negative and ignoring ally damage and deaths.
> 
> ```python
> if self.reward_only_positive:
>     reward = max(0, delta_enemy + delta_deaths)  # shield regeneration
> else:
>     reward = delta_enemy + delta_deaths - delta_ally
> ```

Yes! And this situation happens especially in Maps with Protoss, which can regenerate shields. And I found in the map "3s5z_vs_3s6z" that the allies can learn a strategy to increase the reward without winning the game. Specifically, the allies can learn a pattern to injure the enemies a little and immediately run away from them by hiding in a corner and waiting the enemies to recover, and then repeat. However, the problem can be solved when modifying the reward function to this:

if self.reward_only_positive:
    reward = max(0, delta_enemy + delta_deaths)  # shield regeneration
else:
    reward = delta_enemy + delta_deaths - delta_ally

**Comment by douglasrizzo** [comment]
I sent a quick PR with this tiny change. Check if it solves the issues you have in your experiments.

**Comment by hepengli** [comment]
Thanks! I think this will do. I will check this again and get back to you soon.

**Comment by samvelyan** [comment]
Thanks both for pointing this out and for sending the PR. We're going to see how to best integrate this fix in the upcoming SMAC versions to avoid confusion. One issue we've noticed is that some people compare results using different SMAC/StarCraftII versions and report unfair comparisons between methods in their work.

**Comment by douglasrizzo** [comment]
I have just watched a similar behavior happen in maps which have Medivacs (MMM and MMM2). They have a healing power and my agents have learned to wait for the Medivac to heal the last unit before killing it, sometimes at the expense of the match.

**Comment by hepengli** [comment]
Yes! I've also noticed this situation happens in MMM and MMM2. But your solution by changing the only positive reward to "reward = max(0, delta_enemy + delta_deaths)" is able to fix this issue.

**Comment by xihuai18** [comment]
Why not merge #76 ?

**Comment by samvelyan** [comment]
Given how much the benchmark has been used by the community, fixing this issue now will result in unfair comparisons with existing work. Therefore, we will not merge it with the main branch in this repo.

If you really want to use that particular version of SMAC, you are welcome to use the branch of https://github.com/oxwhirl/smac/pull/76: https://github.com/douglasrizzo/smac/tree/patch-1. But you must make it clear that this is not the standard version of the benchmark when presenting those results.

Lastly, this issue is resolved in SMACv2, the second version of the benchmark which I encourage you to use instead: https://github.com/oxwhirl/smacv2.
